src/federation.cpp

```diff
@@ -467,4 +520,65 @@ void ThreadBlockSigner(NodeContext& node)
     }
 }
 
+void HandleBlockSignMessage(CNode& from, CDataStream& msg, std::chrono::microseconds time_received, const std::atomic<bool>& interrupt_msg_proc)
+{
+    unsigned char code;
+    msg >> code;
+
+    switch (code) {
+        case BlockSignMsgType::VER:
+        case BlockSignMsgType::VERACK:
+        {
+            LOCK(cs_block_signer);
+
+            CPubKey other;
+            msg >> other;
+            if (!other.IsFullyValid()) {
+                LogPrint(BCLog::FEDERATION, "Received blocksign::VER%s message with invalid/malformed public key.  Ignoring.\n", (code == BlockSignMsgType::VERACK) ? "ACK" : "");
+                break;
+            }
+            if (std::find(g_block_signing_keys.begin(), g_block_signing_keys.end(), other) == g_block_signing_keys.end()) {
+                std::string pkstr(HexStr(CDataStream(SER_NETWORK, CLIENT_VERSION) << other), 2);
+                LogPrint(BCLog::FEDERATION, "Received blocksign::VER%s message with unrecognized public key %s.  Ignoring.\n", (code == BlockSignMsgType::VERACK) ? "ACK" : "", pkstr);
+                break;
+            }
+            if (!g_block_signers.count(other)) {
+                std::string pkstr(HexStr(CDataStream(SER_NETWORK, CLIENT_VERSION) << other), 2);
+                LogPrint(BCLog::FEDERATION, "Received blocksign::VER%s message with known public key, but cannot find record for peer: %s.  Ignoring.  Are you missing a a '-blocksignnode=%s:<ip:port>' configuration parameter?\n", (code == BlockSignMsgType::VERACK) ? "ACK" : "", pkstr, pkstr);
+                break;
+            }
+
+            g_fednode_conn[from.GetId()] = other;
+            g_block_signers[other].m_nodes.insert(from.GetId());
+
+            if (code == BlockSignMsgType::VERACK) {
+                break;
+            }
+
+            // FIXME: Don't duplicate HandshakePeer
```

Comment:
bf079d15a6b1581191235f23bc2b53e1131c4034

Maybe just drop this into an inline function above right away. Duplicate code can be very bug-inducing.

Reply:
Done! Please review commit 45a9dbdafa2c239fc13d4bebb79fc6c2f52e8cca

Reply:
LGTM